### ISSUE TITLE
Bulk record requests in a loop cause `ACK_TIMEOUT`s

### ISSUE THREAD

**Issue by Iiridayn** [deepstreamIO/deepstream.io-client-js#270]
Increasing the duration for an `ACK_TIMEOUT` is not a solution when the number of records which must be requested is arbitrary and large - `ACK_TIMEOUT` would have to be arbitrary and large, significantly limiting its value. Instead, please document somewhere a best practice for requesting all records from ids in a list of type `client.record.getList`. Because JS is single threaded, most looping mechanisms either starve the IO which creates `ACK_TIMEOUT`s for records already returned from the server, or are comparatively unacceptably slow (ie, `setTimeout(.., 0)` "looping").

**Comment by Iiridayn** [comment]
This worked unexpectedly well:
```
(function() {
	var requests = [];
	var timeout;
	var runtime = 10; // only run for 10ms max
	var delay = 0;

	function processQueue() {
		var start = new Date().getTime();
		for (var i = 0, l = requests.length; i < l; i++) {
			var since = new Date() - start;
			if (since >= runtime)
				break;
			requests.shift()(); // call the callback
		}

		if (requests.length) {
			timeout = window.setTimeout(processQueue, delay);
		} else {
			timeout = null;
		}
	}

	window.dsRequest = function(callback) {
		requests.push(callback);
		if (!timeout)
			timeout = window.setTimeout(processQueue, delay);
	};
})();
```

It would be called something like:
```
client.record.getList('foo').whenReady(list => {
	list.get().map(id => window.dsRequest(() => client.record.getRecord(id)));
})
```

I also tried `100` and `50` for `runtime` but neither worked. Could probably be tuned, but works well enough for now.

**Comment by ronag** [comment]
@yasserf: We have a bit of similar issues. Maybe one idea is to reset the ack timeout when ANY data is received from the server or something along those lines?

**Comment by Iiridayn** [comment]
```
(function() {
	var cbq = []; // just a single queue, since just a single thread in JS
	var cbqTimeout;
	var cbqRuntime = 8; // only run for 8ms max
	var cbqDelay = 0;
	var cbqActive = 100; // TODO: tune this number
	function processCallbackQueue() {
		var start = new Date().getTime();
		for (var i = 0, l = Math.min(cbqActive, cbq.length); i < l; i++) {
			var since = new Date() - start;
			if (since >= cbqRuntime)
				break;

			cbqActive--;
			// TODO: cut off processed section of the array after, don't shift
			// each iteration of the loop. #microoptimization
			cbq.shift()(function() { cbqActive++; }); // call the callback
		}

		if (cbq.length) {
			cbqTimeout = window.setTimeout(processCallbackQueue, cbqDelay);
		} else {
			cbqTimeout = null;
		}
	}
	window.callbackQueue = function(callback) {
		cbq.push(callback);
		if (!cbqTimeout)
			cbqTimeout = window.setTimeout(processCallbackQueue, cbqDelay);
	};
})();
```
Improved. My setup work once I have a record can be moderately expensive, so I reduce how many can be in flight at once. The final step would be automatically tuning based on the client's capacity and re-requesting records for which there was still an `ACK_TIMEOUT`.

**Comment by yasserf** [comment]
How many subscriptions do you do in one loop?

**Comment by layanto** [comment]
Can ds-queue or async.queue be used to handle concurrency of bulk record requests?

**Comment by Iiridayn** [comment]
@yasserf The biggest currently in the browser (where we're running this) is about 1000 all at once - however, we have some moderatelyl expensive setup work, involving creating an object with knockout computed observables for each field in the schema. With my computer/browser/source code efficiency/etc it seems to print error messages for anywhere from 80 to 800 records. Those 1000 are also preceded by some highly variable number, from about 10 (mostly blank screen) to potentially hundreds (there can be many records, each pulling in display dependencies, which can occasionally be nested).

@layanto I assume if they support a similar interface they should work. I've not looked at third party libraries yet - usually our problem is small enough and simple enough to make it hard to justify adding a dependency to the the project (especially when some common node modules pull in hundreds of others as their own dependencies). We are also running this all in the client, without a CommonJS require system (hasn't been a priority at this stage of development), so node modules wouldn't work well for us on the client anyway :).

**Comment by layanto** [comment]
I haven't tried myself (but will give it a go), d3-queue can be used in the browser and is only 700bytes gzipped.

**Comment by Iiridayn** [comment]
That is a good interface. The code isn't bad either, despite optimizing for
size. Ultimately an async queue doesn't prevent all ACK_TIMEOUTs - just
makes them less likely, depending on tuning and system performance. One can
always perform too much work and starve the network read portion of the JS
loop, till after the timeout, since the timeout portion runs first, even
with a single record request.

My current plan is to implement the async queue (done enough for now), then
a deepstream error handler which catches ACK_TIMEOUTs and re-requests the
records (somehow preserving original record callbacks, etc), up to some
arbitrary but small count, such as 3 times, before finally throwing an
error and giving up. I suspect this would make the client much more robust.

2017-01-06 18:00 GMT-07:00 Jeffrey Layanto <notifications@github.com>:

**Comment by slachtar** [comment]
Hi,
do anyone already found a clean solution for ack timeouts? We struggle because of that and results are arbitrary because not all records are subscribed.
Regards.

**Comment by yasserf** [comment]
Hey All,

I'm keen to get a solution for timeouts out and have been thinking alot about possible solutions.

The problem first of isn't that acks indicate something doesn't work, it just means that its taking longer than possible. This could be due to a large amount of subscriptions ( thousands simultaneously ) occurring at the same time on the server ( we are working on getting exact number ) but does NOT mean your subscription failed. 

The only time a subscription would fail is if the server dies, but we have normal failover scenarios for that, mainly that as soon as the connection is reestablished they are all resubscribed normally.

So the problem is actually the fact ack timeouts happen, which is either due to:

- No free cpu cycles on the client where the library doesn't have time to get something off the wire
- No free cpu cycles on the server meaning it takes longer to subscribe subscriptions  (this occurs very rarely)

As such, I suggest that use a slightly different approach, where we push ack timeouts forward when we know that the results would take longer, so all timeouts fire once the last timeout occurs.

This might seem odd, but if your subscribing to lots of records in a loop javascript event loop starves.

Thoughts?

In summary:

subscribe to a ( timeout a for 20 ms )
10ms pass
subscribe to b ( timeout for a and b in 20ms )
subscription a ack comes in
subscribe to c ( timeout for b and c in 20ms )

**Comment by WolframHempel** [comment]
this probably ties in with the introduction of a `warning` event to avoid ack-timeouts / unsolicit messages being perceived as errors

**Comment by ronag** [comment]
I don't think having a timeout in its current form makes any sense. Isn't it enough to just have one single timeout, i.e. the client is waiting for a (any) response from the server? I'm not sure what the advantage of these fine-grained timeouts is? i.e. if the client haven't lost connection and the server is bug free there should always be a response...

**Comment by yasserf** [comment]
Acks are used to indicate things are working as efficiently as possible. While they don't indicate any sort of actual error ( as @WolframHempel stated ) if your seeing them occur it means you either have seriously bad latency or your application is not working smoothly. If you subscribe to something and don't get a response within a few hundred milliseconds ( which is quite high ) it usually means you are reserving the cpu for too long.

If you don't benefit from knowing there are bottlenecks in your code ( or that you might be potentially putting the server under ALOT of load ) or you intentionally have expensive threads then you can usually compensate by just setting the timeout to a much larger number.

What we will be fixing soon is all the odd scenarios around timeouts, like when you subscribe before login -> shouldn't start tracking, and if you lose your connection it should clear all remaining timeouts.

The warning event would make it alot less noisy, so that you'll only get told about it if you actively listen.

**Comment by ronag** [comment]
@yasserf: And what is the advantage of fine grained timeouts? Your scenario would work as well with a single global timeout.

**Comment by ronag** [comment]
```js
send(msg) {
  if (this.pending++ === 0) {
     this.timeout = setTimeout(...)
  }
  ...
}

ack () {
  this.pending--
  clearTimeout(this.timeout)
  this.timeout = this.pending > 0 ? setTimeout(...) : null
  ...
}
```

**Comment by yasserf** [comment]
If I keep sending things that reset a global timeout on a loop it would just never fire. That is not expected behavior.

The advantage of fine grained timeout topics/subjects is that it gives a bit more context on what starved or broke when. We can easily remove having separate timeouts with a global one, but I would still want it to to warn about all the subscriptions that got delayed because 

- its part of the protocol and would be a waste if we didn't
- it helps quite a bit with debugging

Once it starts notifying on warn rather than error you can just ignore any timeouts, and we'll try to remove multiple timeout functions and replace them with an interval to improve efficiency by reducing time/memory allocated to the timeout registry when too many timeouts occur.

**Comment by ronag** [comment]
> If I keep sending things that reset a global timeout on a loop it would just never fire. That is not expected behavior.

No. You only reset the timer when you receive an ack. Basically you want if I have pending acks and have not received ANY ack in a time period, THEN we have a timeout.

> The advantage of fine grained timeout topics/subjects is that it gives a bit more context on what starved or broke when.

Does it? If one thing starves then everything starves, no? I'm not sure how useful the context is?

> its part of the protocol and would be a waste if we didn't

feature vs performance and quality...

> it helps quite a bit with debugging

How?

**Comment by slachtar** [comment]
Sorry, I'm getting a little bit confused and we need to put our app in production very soon.
Dear @yasserf should we ignore ACK_TIMEOUT and READ_TIMEOUT errors if they raise in the client side and not in the server side because the subscriptions did really happen.
BTW we noticed that getting records and subscribing in a loop leads to less errors if we introduce a delay of 100ms via setTimeout in each step of the loop.
Or should we listen to client errors and re-get and re-subscribe records with errors.
Regards

**Comment by yasserf** [comment]
I think we just have two separate schools of thought. We try to avoid subscribing to large amounts of things in one go upfront, and that way we can keep spikes on our event loop to a minimum and would like to be told when performance is below what we expect. Whenever I see an ack timeout occur on any deepstream application I investigate why and usually find a bottleneck we can overcome.

As such I doubt we will reach an agreement on a the usefulness they provide. What we do will reduce its visibility and impact significantly by moving it to an opt in warn status, and potentially put in the ability to disable them entirely if you don't feel they aid you in anyway.

I'll update this issue once we have a release planned, but in summary:

- clear timeouts on disconnect
- dont create timeouts before authentication
- move non fatal messages to a warn instead of a error
- investigate possible migration to a single timeout for performance improvements
- **potentially** allow timeouts to be disabled.

**Comment by yasserf** [comment]
@slachtar yes, you can ignore ack timeouts and read timeouts without any side effects. If any actual error occurs with will either be a CACHE or STORAGE error

You can't resubscribe to something when an ack timeout occurs because deepstream will already have your previous subscription either in the queue or processed, leading to a MULTIPLE_SUBSCRIPTION warning, so that shouldn't be an option!

**Comment by slachtar** [comment]
@yasserf thank you for the clarifications.

**Comment by layanto** [comment]
I use d3-queue to limit the number of getRecord and subscribe at any one time. I haven't had to retrieve lots of records yet (but will need to later on) so not sure if this approach will help. I use d3-queue to handle getRecord and record.subscribe but calling record.discard in loop directly. Not sure how to handle record.discard with d3-queue as there is no callback in discard. With the changes proposed by @yasserf, do I need to do anything differently?

Code sample below (in vuejs) - still trying to see if I can write the code better as currently a lot of repetitive boilerplate in each vuejs component:
```javascript
<template>
    <div>
        <table class="responsive-table striped">
            <thead>
                <tr>
                    <th data-field="username">Username</th>
                    <th data-field="firstname">First Name</th>
                    <th data-field="lastname">Last Name</th>
                </tr>
            </thead>
            <tbody>
                <tr v-for="user in sortedUsers">
                    <td>{{ user.username }}</td>
                    <td>{{ user.firstname }}</td>
                    <td>{{ user.lastname }}</td>
                </tr>
            </tbody>
        </table>
    </div>
</template>

<script>
import { queue } from 'd3-queue'
import _ from 'lodash'
export default {
    data () {
        return {
            list: '',
            listEntries: [],
            records: [],
            users: []
        }
    },
    computed: {
        sortedUsers: function () {
            return this.users.sort(function (a, b) {
                if (a.username < b.username) return -1
                if (a.username > b.username) return 1
                return 0
            })
        }
    },
    methods: {
        getList: function () {
            const queryString = JSON.stringify({
                table: 'user',
                query: true
            })
            this.list = this.$store.state.ds.client.record.getList('search?' + queryString)
            this.list.whenReady((list) => {
                list.subscribe((entries) => {
                    const removed = _.difference(this.listEntries, entries)
                    const added = _.difference(entries, this.listEntries)
                    this.listEntries = entries
                    const qr = queue(20)
                    added.forEach((id) => {
                        qr.defer(this.getRecord.bind(this), 'user/' + id)
                    })
                    qr.awaitAll((err, data) => {
                        if (err) {
                            console.log(err)
                        } else {
                            this.records = _.concat(this.records, data)
                            const qs = queue(20)
                            data.forEach((record) => {
                                qs.defer(this.subscribeRecord.bind(this), record, this.users)
                            })
                            qs.awaitAll((err) => {
                                if (err) console.log(err)
                            })
                        }
                    })
                    removed.forEach((id) => {
                        const records = _.remove(this.records, (n) => {
                            return n.name === 'user/' + id
                        })
                        records.forEach((record) => {
                            record.discard()
                        })
                        this.users.splice(_.findIndex(this.users, (n) => {
                            return n.id === 'user/' + id
                        }), 1)
                    })
                }, true)
            })
        },
        getRecord: function (recordName, callback) {
            var record = this.$store.state.ds.client.record.getRecord(recordName)
            record.whenReady((record) => callback(null, record))
        },
        subscribeRecord: function (record, arr, callback) {
            record.subscribe((data) => {
                data.id = record.name
                const index = _.findIndex(arr, (n) => {
                    return n.id === data.id
                })
                if (index > -1) {
                    arr.splice(index, 1, data)
                } else {
                    arr.push(data)
                }
                callback(null)
            }, true)
        }
    },
    mounted: function () {
        this.getList()
    },
    destroyed: function () {
        this.list.discard()
        this.records.forEach((record) => {
            record.discard()
        })
    }
}
</script>
```

**Comment by ronag** [comment]
This discussion is actually making me consider disabling ACK on the server for "production" mode.

**Comment by ronag** [comment]
> we can keep spikes on our event loop to a minimum

I think the best tool for that would be to use `requestIdleCallback`...

**Comment by yasserf** [comment]
Fixed in 4.0 by putting all requests in one message (bulk subscribe) and getting a single ACK back with a correlation id. It returns an ack on the first message it processes and not the last so that you get a response as soon as possible. I done it with 40k messages and it worked fine!
